asahi_firmware/multitouch.py

```diff
@@ -29,7 +29,59 @@ def unmunge(el, idmap):
 
     return plistlib.loads(ET.tostring(pl))
 
-def plist_to_bin(plist):
+LOAD_COMMAND_INIT_PAYLOAD = 0
+LOAD_COMMAND_SEND_BLOB = 1
+LOAD_COMMAND_SEND_PROPERTY = 2
+
+def plist_to_bin_touchscreen(plist):
+    def swap32(i):
+        a = (i >> 8) & 0xFF
+        b = i & 0xFF
+        c = (i >> 24) & 0xFF
+        d = (i >> 16) & 0xFF
+        return a << 24 | b << 16 | c << 8 | d
+    def serialize(plist):
+        for i in plist:
+            if i["Type"] == "Config":
+                init = i["Config"]["SPI Config"]["Init Payload"]
+                yield struct.pack("<II", LOAD_COMMAND_INIT_PAYLOAD, len(init))
+                yield init
+                realign = len(init) % 4
+                if realign != 0:
+                    yield bytes([0] * (4 - realign))
+        for i in plist:
+            pkt_len = 0
+            if i['Type'] == 'Binary':
+                payload = i['Payload']
+                pkt_len = 14 + len(payload)
+                yield struct.pack("<II", LOAD_COMMAND_SEND_BLOB, pkt_len)
+                addr = i['Address']
+                hdr = struct.pack(">HHI", 0x3001, (len(payload) + 3) // 4, swap32(addr))
+                yield hdr
+                yield struct.pack(">H", sum(hdr[2:]))
+                for i in range(len(payload) // 4):
+                    yield bytes([payload[i * 4 + 1], payload[i * 4], payload[i * 4 + 3], payload[i * 4 + 2]])
+                checksum = sum(payload)
+                yield struct.pack(">I", swap32(checksum))
+            elif i['Type'] == 'Property':
+                yield struct.pack("<II", LOAD_COMMAND_SEND_PROPERTY, i['Address'])
+            elif i['Type'] == 'RequestCalibration':
+                pkt_len = 2
+                yield struct.pack("<II", LOAD_COMMAND_SEND_BLOB, pkt_len)
+                yield struct.pack(">H", 0x1f01)
+            elif i['Type'] == 'ReadModifyWrite':
+                pkt_len = 16
+                yield struct.pack("<II", LOAD_COMMAND_SEND_BLOB, pkt_len)
+                data = struct.pack(">HIII", 0x1e33, swap32(i['Address']), swap32(i['Mask']), swap32(i['Value']))
+                yield data
+                yield struct.pack(">H", sum(data[2:]))
+            realign = pkt_len % 4
+            if realign != 0:
+                yield bytes([0] * (4 - realign))
+    return b"".join(serialize(plist))
```

Comment:
I would add a header here (note how I have the HIDF one below), just so we have some ability to add new versions/formats.

Maybe `hdr = struct.pack("<4sIII", b"Z2FW", 1, 16, len(blob))` (where 1 is the version and 16 the offset of the blob relative to the header).